lib/features/modeling/cmd/ReconnectConnectionHandler.js

```diff
@@ -29,38 +27,36 @@ ReconnectConnectionHandler.prototype.execute = function(context) {
 
   context.oldWaypoints = oldWaypoints;
 
-  if (isArray(dockingOrPoints)) {
-    newWaypoints = dockingOrPoints;
-  } else {
-    newWaypoints = oldWaypoints.slice();
-
-    newWaypoints.splice(newSource ? 0 : -1, 1, dockingOrPoints);
-  }
-
   if (newSource) {
     context.oldSource = connection.source;
     connection.source = newSource;
-  }
-
-  if (newTarget) {
+  } else if (newTarget) {
     context.oldTarget = connection.target;
     connection.target = newTarget;
   }
 
-  connection.waypoints = newWaypoints;
-
   return connection;
 };
 
 ReconnectConnectionHandler.prototype.postExecute = function(context) {
   var connection = context.connection,
-      connectionStart = getDocking(connection.waypoints[0]),
-      connectionEnd = getDocking(connection.waypoints[connection.waypoints.length - 1]);
+      dockingOrPoints = context.dockingOrPoints,
+      newSource = context.newSource,
+      movedEnd = newSource ? 'connectionStart' : 'connectionEnd',
+      newWaypoint,
+      hints = {};
+
+  if (isArray(dockingOrPoints)) {
+    connection.waypoints = dockingOrPoints;
```

Comment:
We cannot leak this side effect in post execute. We must handle it (and properly revert it) in execute.